### ISSUE TITLE
qBittorrent port not updating

### ISSUE THREAD

**Issue by markus-lei** [SnoringDragon/gluetun-qbittorrent-port-manager#5]
On my system, the container is logging that it successfully updated the qBittorrent port, but it does not change.

I am unsure how to troubleshoot this (since the log says it is successful)..?

**Comment by JakubKopys** [comment]
On which port is you qBittorrent web ui running? The PORT in `QBITTORRENT_PORT` has to correspond to the web ui port, so if you change your qbittorent ui with:
```
    environment:
      - WEBUI_PORT=6011
```

then you have to set the `QBITTORRENT_PORT` to 6011 too.

**Comment by markus-lei** [comment]
Hi, 

my qBittorrent is on port 8080, I have configured this and it is also reachable there. I have also double checked that no other container is using this port.

**Comment by ErnToTo** [comment]
The bash script inside this repo works by monitoring changes in gluetun-generated file with forwarded port. If it somehow misses that info (container restarts, qbitorrent restarts etc) it will not update the port. This PR  #7 fixes the issue by setting timeout value.

**Comment by markus-lei** [comment]
I understand, but according to the log he finds it just fine... Problem is, even after he logs that he successfully updated it in qBittorrent - it is not actually changed there.

**Comment by whatfreudsaid** [comment]
I'm having the same problem. The container logs that the port has been changed, but it remains unchanged when i check to see inside the qBittorrent Web UI.

It is not a problem of finding the port (in the gluetun generated file), because the reported port is the correct one. It seems to be the problem of actually changing the port inside qBittorrent.

**Comment by DlNGDlNG** [comment]
Had this exact issue where logs said port was updated but they did not. In this section of your compose file:

environment:
      QBITTORRENT_SERVER: localhost

have you entered localhost or the IP address of your qBittorrent instance?

**Comment by 5at0ri** [comment]
I've got this issue, when I restart the container for this port manager using Portainer is updates the port, but doesn't do it automatically. It seems it doesn't automatically detect the changes but I'm not sure

**Comment by kennyp87** [comment]
> I've got this issue, when I restart the container for this port manager using Portainer is updates the port, but doesn't do it automatically. It seems it doesn't automatically detect the changes but I'm not sure

I've got exactly the same issue too. In the log files for gluetun I can see that the connection has changed, and a new forwarded port is updated to the file. But it does not update automatically, only if I restart the container for this project.

**Comment by 5at0ri** [comment]
> > I've got this issue, when I restart the container for this port manager using Portainer is updates the port, but doesn't do it automatically. It seems it doesn't automatically detect the changes but I'm not sure
> 
> 
> 
> I've got exactly the same issue too. In the log files for gluetun I can see that the connection has changed, and a new forwarded port is updated to the file. But it does not update automatically, only if I restart the container for this project.

Yup, looks like it's the same issue. Hopefully it can get fixed.

**Comment by RMerl** [comment]
The inotify wait doesn't seem to be working as expected.  I had the same issue where the VPN got reconnected, the forwarded_port content was updated, but this docker never updated the port.  Starting a shell within the container and killing the currently running inotify_wait properly updated the port in qbittorent, confirming that the issue was with that, not with the communication between this docker and qbittorrent.

I think inotify can be tricky with mounted volumes with docker, but I don't know enough about docker to dive any deeper.

**Comment by 5at0ri** [comment]
just wanted to share that I have a janky ish solution to this. since the port gets updated when the container restarts I tried to use Ofelia to make it work but im not too great with docker so I couldn't get it to work. however I have just successfully set up a cron job to restart the container every minute, therefore making the port get updated every minute ensuring the qbit won't get firewalled if the port changes. My container name for it is compose-gluetun-qbittorrent-port-manager-1, and I've put exactly what I did below. I confirmed this is working by monitoring the start time of the container in portainer and it seems to be working!

sudo crontab -e

* * * * * /usr/bin/docker restart compose-gluetun-qbittorrent-port-manager-1 >/dev/null 2>&1       #add this to the end of the thing (I used nano which was the first option for me)

verify it saved with sudo crontab -l

Hope this helps anyone

**Comment by 3r1c02** [comment]
I've had the same problem.

The solution was changing environment variables that the script uses directly
```
env:
- name: QBITTORRENT_SERVER
  value: "localhost"
- name: QBITTORRENT_PORT
  value: "8080"
- name: "HTTTP_S"
  value: "http"
- name: QBITTORRENT_USER
  value: "admin"
- name: QBITTORRENT_PASS
  value: "hidden"
```

The start.sh uses these to change the port in qbittorrent with these curl request

```
  curl -s -c $COOKIES --data "username=$QBITTORRENT_USER&password=$QBITTORRENT_PASS" ${HTTP_S}://${QBITTORRENT_SERVER}:${QBITTORRENT_PORT}/api/v2/auth/login > /dev/null
  curl -s -b $COOKIES --data 'json={"listen_port": "'"$PORT"'"}' ${HTTP_S}://${QBITTORRENT_SERVER}:${QBITTORRENT_PORT}/api/v2/app/setPreferences > /dev/null
```

**Comment by Noggog** [comment]
For me, if I modified the gluetun file directly after everything started, it updated properly.   This suggests that perhaps it's a startup race condition?  It could be that this docker updates the port, which is then replaced with the saved settings of qbittorrent as it's starting up.

**Comment by huuaho** [comment]
I had this issue as well - the docker logs said that it updated but querying the client showed that the port was not updated,

I made a fork adding some CSRF headers and that worked for me

**Comment by dkman123** [comment]
I agree with Noggog.  I think in my case that port manager is reading the old file and setting the port, then gluetun finishes starting and updates the file with the new port.
If the 10 second sleep was customizable I would change it to 60, or even 90, so that gluetun had a chance to finish, then portmanager would read the correct port number.

Another option is to "docker restart qbitportman" after starting everything starts (replace with your container name). Even though I have a depends on gluetun service_healthy it fires off too soon.  Or my port is changing at some later point and not getting picked up.

I'm considering a scheduled job to restart port manager every 6 hours just to make sure, since it's light enough that it wouldn't hurt anything.

**Comment by Amekai777** [comment]
I had the same issue 
Port Manager was saying it has successfully updated (But was not true - Remained firewalled and old port number in UI) 
I wonder if the attempt was timing out and being reproted as successful 

I was able to solve this with trial and error .. Not sure which bit worked, but to help others :-
1) I had originally put the local address in rather than Localhost - CHANGED TO "LOCALHOST"
2) Selected "Bypass authentication for clients on localhost" option within Qbittorrent settings for WebUI
3) I did leave username and password in the Docker Compose for Port manager TBF. 

It is worth noting re the startup timing issues described above 
That my port manager does have 
  depends_on:
      gluetun:
        condition: service_healthy
      qbittorrent:
        condition: service_healthy

So that it's last to activate on stasck initiation

For the first time this correctly updated the port number and confirmed within qbittorrent in UI.

**Comment by blue-belle** [comment]
Having the same issue, if I manually enter a port into `forwarded_port` Qbittorrent's UI reports no change whatsoever
